### ISSUE TITLE
Transforming <> include into ""

### ISSUE THREAD

**Issue by mbrucher** [include-what-you-use/include-what-you-use#496]
Hi,
I'm trying IWYU again on a big code base, and I just changed there all relative includes to absolute ones with <>. But it seems that IWYU wants to replace my <> includes with "" includes even if they are not relative, and even then sometimes replacing some absolute includes with relative ones even if doesn't make sense.
Is tehre a way to force IWYU NOT to change the paths in includes? There is a reason why the paths are absolute.
Also, seems like on a database, it doesn't change includes in headers?

**Comment by kimgr** [comment]
The meaning of `<>` vs `""` is under-specified at best, and IWYU uses some arbitrary interpretation. I think its policy is --

- If a header is found via `-I` it is included with `""`
- If a header is found via `-isystem` it is included with `<>`

So by mucking around with your build system you can trick IWYU into better behavior, but it probably won't behave exactly like you want.

> Also, seems like on a database, it doesn't change includes in headers?

What do you mean by "on a database"?

**Comment by mbrucher** [comment]
Sorry, database was a bad expression for a command.json file from CMake.
The policy doesn't make any sense. "" and <> have a specific meaning that has nothing to do with system, or not system (so yes, there is some kind of specification). "" indicate that you look first in the current folder of the file to find the header and then folders given by -I and -system. In both cases, if the file is in -I and -system it must be <>. And only "" if the file is relative to the current file.

**Comment by mbrucher** [comment]
Seems like the standard is indeed quite fuzzy, but there is no mention of a difference between -I and -system and no compiler that I know of makes such a difference: https://stackoverflow.com/questions/21593/what-is-the-difference-between-include-filename-and-include-filename
GCC considers -I AND -system to be system files in that regard: http://gcc.gnu.org/onlinedocs/cpp/Include-Syntax.html
VS does the same: https://msdn.microsoft.com/en-us/library/36k2cdd4(v=vs.140).aspx
So not sure what this arbitrary interpretation comes from, but it slows down compilation. And using IWYU is about faster build time, so this is definitely not the best way of including files that are in the "include paths".

**Comment by kimgr** [comment]
I think this could be something that fell out of Google style, as IWYU originated in Google. It's not spelled out in the Google style guide, but examples -- e.g. here: https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes -- indicate that they use angle-bracket includes for system headers and quotes for everything else.

It might be a good idea to revisit this, I guess you're saying that `""` should only be used for includes relative to the current file, and `<>` for everything else?

One problem is that the meaning of "current file" has traditionally been different between GCC and MSVC (for GCC it's the current compilation unit, for MSVC it's the including file, whether header or .cpp file being compiled). I don't know if that's still the case now.

**Comment by mbrucher** [comment]
Indeed, there is a differnece, but at least if the file is in the include path (-I or -system for gcc/clang or /I for VS), the solution is easy: use <>.

**Comment by kimgr** [comment]
I don't think that policy is going to be popular with very many code bases even if I agree it has merits.

But this is an interesting angle, it would be nice if IWYU could allow more configurability for `<>` vs `""`. Not sure how many different policies there can be in the world.

**Comment by mbrucher** [comment]
Well, on all code bases I worked on, this was the official policy (and I worked for different companies, R&D departments, commercial apps). Anyway, yes, it would be nice, as the <> policy is actually faster than "", and it can make a noticable difference on some codebases/fiesystems.

**Comment by ThosRTanner** [comment]
The official language definition is that

#include <...> looks for the include file in a compiler specified place.

#include "..." looks for the include file in a different compiler specified place
if it can't find it there, it treats the line as if you'd written #include <file>

using "..." does not mean that the compiler will look in the current directory of the including file first, at least not on all compilers.

I've used compilers where specifying a user specified file in <> will cause the compilation to fail because the file can't be found.

I think really you need a way of telling IWYU how you want it to behave. For the places where I've worked, the current rules (is per @kimgr 's first post) seem pretty applicable.

**Comment by mbrucher** [comment]
Which compiler was it? Is it a compiler that IWYU "supports"?

**Comment by ThosRTanner** [comment]
I thought IWYU supported the language, not the compiler.

**Comment by kimgr** [comment]
Well put. That is generally the intent, yes. But we try to accommodate "the real world" from time to time :)

**Comment by mbrucher** [comment]
If the support is for the language, then there is a lack for selecting <> or "".
The reality is still that you need to have clang being able to compile the code, so everything will be biased towards clang by definition.
Not even talking about the fact that the header usage is heavily biased towards gcc/clang headers!

**Comment by ThosRTanner** [comment]
Some of us have to cross compile. In any case I agree that some sort of customisation would be nice.

And in passing, IWYU doesn't really seem to deal with the `#include "..."` starting in the directory of the file containing the `#include` line very well (or at least not the version I have) and puts in the whole path.

**Comment by chri7325** [comment]
For our codebase, we lock things down to include with <> due to the predictable nature of it.  From what I remember, I think including with "" has behavior that is compiler specific.  Since we use three separate compilers (msvc, clang, gcc), we want to limit this difference.  Our coding standards are based on http://www.stroustrup.com/JSF-AV-rules.pdf which specifies this under the header section.  We are about to use iwyu on our entire codebase (by the way, thank you so much for this amazing tool; it has reduced out compile times by over 66% in our prototype) and I've needed to make a script that will automatically change include " to include <.  It isn't the worst thing in the world but it would be great to get the behavior we're looking for out of the box through some kind of option.

**Comment by kimgr** [comment]
@chri7325 The patch looks nice and clean! I'm generally not very forthcoming with new command-line arguments; we really want IWYU to just do the right thing. But this appears to be an axis where everybody has their own truth.

Is it possible to generalize this somewhat? So not only `--angled_includes`, but something like `--include_style` assuming we can make up names for all the weird policies people use?

**Comment by chri7325** [comment]
@kimgr sounds good to me.  Would it need something like --include_style=<angled|quoted(default)> and --include_style_system=<angled(default)|quoted>?

**Comment by chri7325** [comment]
@kimgr In our use-case, we don't allow quoted includes anywhere so it was easy to just specify angled includes for everything.  For private project includes, we only expose the project root on the include path for that project.  From there, internal includes are added using the path to the include from src.  This makes it easy to immediately identify that an include is private as well as keeping all the private includes grouped since we use clang-format to sort them.  Would IWYU respect that if we're only adding include_style to normal includes?  We had an issue where iwyu would read an include with <>, suggest keeping it in the log as "" and then not realize that it was already included with <> so we had to add <> to every include so it was able to properly identify that the include was already added.  I just hope isolating the cases to only normal includes respects that.
```
#include <other_project/public_header.h>
#include <same_project/public_header.h
#include <src/private_include_that_isnt_exposed_to_other_projects.h>
```

**Comment by horenmar** [comment]
+1 on having this issue. We use full path from project root in `<>` for includes, and having IWYU suggest removing a path with `<>` but adding the exact same path `""` is completely useless.

**Comment by spectras** [comment]
In my company, the policy is `<>` for 3rd party includes and `""` for home-grown includes.

It has no connection to path begin relative or absolute whatsoever.

Thus configurability would be great. On the other hand trying to steer iwyu into enforcing your local style not so much.

**Comment by kimgr** [comment]
@spectras Yes, I agree it would be nice to have this configurable. But I'm not sure what the configuration matrix looks like. I haven't thought about this in a while, but last time I did, I dug myself into a pretty complex hole :).

Do you have any ideas for how you'd like to express this?

**Comment by mpusz** [comment]
I would also love to see it configurable. I am providing a header-only library so my files have to be included with `<...>`. Using `-isystem` is not an option for me because this silences some warnings in a compiler and I want to be aware of any issue with my code and fix it ASAP.

**Comment by johnmcfarlane** [comment]
C++ Core Guideline on the subject in case it helps: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rs-incform

I'm considering trying out this util but not until it reflects how POSIX and MSVC work.

**Comment by kimgr** [comment]
@johnmcfarlane Nice, this is the first mildly authoritative writing I see on the subject. It sounds to me like a quote-style could be passed on the command-line, e.g. `--quote_style=(google|cppcore|always_angled)` or something like that. I think those are the three major styles I've seen (and maybe one additional `whatever`, which is typically the most prevalent :-)). And `cppcore`, in that case, would make a good default.

The only open question is whether it's easy/possible to detect whether includer and includee are locally relative in all the strange cases we conjure up includes.

**Comment by mbrucher** [comment]
The core guidelines basically states that if you can use <> without changing the "argument", then you should use it, and I think that's a good thing. I don't think `always_angled` can work though, without changing something and I'd rather have the user change paths if needed than IWYU. It's very easy to get the wrong include in for whatever reason (CLion does that sometimes, and at least in the codebase I have, it's always wrong).

**Comment by mpusz** [comment]
BTW, see this one for an example: https://github.com/mpusz/units/blob/master/example/avg_speed.cpp.

**Comment by johnmcfarlane** [comment]
@mpusz 

> For example, in my library, I deliberately use <units/....> to refer to all headers as they will be exported in the interface and used by the customers. So you should not check if the source and header file are in the same project (relative to each other) or not.

It sounds like your project would follow the `--quote_style=always_angled`. So the question does not apply to your project.

For `--quote_style=cppcore`, thinking out loud:

```python
if the included file is found under a header search path:
  if the including file is a source file:
    use <>
  else if the including file is a header file:
    if the including file is found under the same header search path as the included file
      use ""
    else
      use <>
else
  use ""
```

Not sure that's correct or complete though. It might be worth thinking about how one might write a Clang-Tidy rule enforcing rule SF.12 and working back from there.

**Comment by mpusz** [comment]
> It sounds like your project would follow the --quote_style=always_angled. So the question does not apply to your project.

Well, not really. It applies only to the "library-part" of the project. It will not work for usage examples:

https://github.com/mpusz/units/blob/309da80c32c9c48c43f3daa6ec8db31d9075a4e5/example/glide_computer_example.cpp#L23-L27

**Comment by mpusz** [comment]
I am not sure about the following:

> ```
> if the including file is found under the same header search path as the included file
>   use ""
> else
>   use <>
> ```

Does this mean that `<vector>` should include `"bits/stl_vector.h"` rather than `<bits/stl_vector.h>`?

**Comment by johnmcfarlane** [comment]
I'm not sure that GCC follows the CCG. If it did, then yes, assuming it's not doing something special (e.g. providing some kind of preprocessor customisation point) then yes, `""` would make it slightly harder for other search paths containing a file called _bits/stl_vector.h_ to break things.

Assuming that

* the standard library was an ordinary library, and
* the header search paths were provided using `-isystem /bogus/library -isystem /usr/include/c++/10`, and 
* there was a file, _/bogus/library/bits/stl_vector.h_

then you can see how using `<>` would break `std::vector`! That's one reason for the guideline. However, that's a rather far-fetched example. I think most implementers pull in their standard library files via a different method and the rules for that library are different to most other libraries. (For starters, `<vector>` doesn't even need to be a file at all.)

**Comment by mpusz** [comment]
What I meant by the above is that:
1. External libraries should be included by dependents with `<>`
2. External libraries should refer to their implementation details with `<>`
3. Those external libraries have their own repositories/projects where they are implemented and tested
  - in those projects, they should not be included with `-isystem` because this disables some compiler warnings which means that some bugs can leak to production
  - their headers should also be included with `<>` at least in the usage examples but probably also in unit tests (for consistency)
  - if there are any headers specific to unit tests or examples (not part of the exported library) they should be included with `""` when provided as a relative path and with `<>` when provided as an "absolute" path.

**Comment by excitoon** [comment]
https://www.stroustrup.com/JSF-AV-rules.pdf
![Screenshot from 2024-04-29 22-58-47](https://github.com/include-what-you-use/include-what-you-use/assets/2255111/8c9161db-46d9-47f8-924a-612b2bed4ab7)

**Comment by excitoon** [comment]
However we shall respect _user_ choice of syntax.

**Comment by justusranvier** [comment]
I just ran into this issue today where iwyu insisted on adding a bracket include when the desired quote form was already present.

Most of the conversation in this thread has been about comprehensive solutions but perhaps there can be an intermediate step where the mapping file could be used to add per-header overrides to handle individual cases where iwyu does the opposite of what the user wants.

```
{ "style": ["foo/bar.hpp", "bracket"] },
{ "style": ["something/else.hpp", "quote"] }

```
